src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs

```diff
@@ -52,59 +58,115 @@ public override async ValueTask<int> ReadAsync(Memory<byte> data, CancellationTo
 #if NETSTANDARD2_0
         var data = buffer.AsMemory(offset, count);
 #endif
+        var headerBuffer = Memory<byte>.Empty;
 
         switch (_state)
         {
             case ResponseState.Ready:
-                // Read the header first
-                // - 1 byte flag for compression
-                // - 4 bytes for the content length
-                Memory<byte> headerBuffer;
-
-                if (data.Length >= 5)
                 {
-                    headerBuffer = data.Slice(0, 5);
+                    if (data.Length == 0)
+                    {
+                        // Handle zero byte reads.
+                        var read = await StreamHelpers.ReadAsync(_inner, data, cancellationToken).ConfigureAwait(false);
+                        Debug.Assert(read == 0);
+                        return 0;
+                    }
+                    else
+                    {
+                        // Read the header first
+                        // - 1 byte flag for compression
+                        // - 4 bytes for the content length
+                        _contentRemaining = HeaderLength;
+                        _state = ResponseState.Header;
+                        goto case ResponseState.Header;
+                    }
                 }
-                else
+            case ResponseState.Header:
                 {
-                    // Should never get here. Client always passes 5 to read the header.
-                    throw new InvalidOperationException("Buffer is not large enough for header");
+                    Debug.Assert(_contentRemaining > 0);
+
+                    headerBuffer = data.Length >= _contentRemaining ? data.Slice(0, _contentRemaining) : data;
+                    var success = await TryReadDataAsync(_inner, headerBuffer, cancellationToken).ConfigureAwait(false);
+                    if (!success)
+                    {
+                        return 0;
+                    }
+
+                    // On first read of header data, check first byte to see if this is a trailer.
+                    if (_contentRemaining == HeaderLength)
+                    {
+                        var compressed = headerBuffer.Span[0];
+                        var isTrailer = IsBitSet(compressed, pos: 7);
+                        if (isTrailer)
+                        {
+                            _state = ResponseState.Trailer;
+                            goto case ResponseState.Trailer;
+                        }
+                    }
+
+                    var read = headerBuffer.Length;
+
+                    // The buffer was less than header length either because this is the first read and the passed in buffer is small,
+                    // or it is an additonal read to finish getting header data.
+                    if (headerBuffer.Length < HeaderLength)
+                    {
+                        _headerBuffer ??= new byte[HeaderLength];
```

Comment:
this `byte[]` troubles me; it is always `5`, right? well, we can get that far more cheaply by abusing a `long` field (which is the same size as the `byte[]` *reference*, but without needing the `byte[]`) - perhaps thunking as a `Span<byte>`:

``` c#
    private long _header;
    private Span<byte> GetHeader() => MemoryMarshal.Cast<long, byte>(new Span<long>(ref _header)).Slice(0, 5);
```

similar can be done on down-level using `unsafe`, or you can just use shifts

Is it worth the evil, though? perhaps:

``` c#
#if NET7_0_OR_GREATER
    private long _header;
    private Span<byte> GetHeader() => MemoryMarshal.Cast<long, byte>(new Span<long>(ref _header)).Slice(0, 5);
#else
    private byte[]? _header;
    private Span<byte> GetHeader() => _header ??= new byte[5];
#endif
```

Reply:
```c#
MemoryMarshal.CreateSpan(ref Unsafe.As<long, byte>(ref _headerBuffer), 5);
```
will likely get better codegen

Reply:
yeah, I thought about that alternative - it does lose the bounds check, but on the other hand, it will never be true that `sizeof(long) < 5`, so that bounds check is redundant... but: what *really* makes that tempting is that `MemoryMarshal.CreateSpan` approach allows us to go all the way back over all netcoreapp2+ and netstandard2.1, so: much better TFM coverage 👍

Reply:
Header buffer is passed to async methods and needs to be `Memory<byte>`.

This allocation only happens if someone uses a small buffer (less than 5 bytes) with the stream. The primary user of the stream is `Grpc.Net.Client`, which uses a much larger buffer.

Realistically, I doubt people will hit this allocation in the real world.

Reply:
Could we just throw `NotSupportedException` or something if someone tries to read fewer than `HeaderCount` bytes?  This seems like a lot of complexity for code we don't expect or want people to use.